src/components/PromotionList.tsx

```diff
@@ -50,6 +55,18 @@ export default function PromotionList({
       });
   }, [promotionListState.filter, promotionListState.sort, dispatch]);
 
+  /**
+   * When a search query is set, fetches promotions that satisfy the query.
+   */
+  useEffect(() => {
+    if (promotionListState.searchQuery) {
+      getPromotions([{ searchQuery: promotionListState.searchQuery }]).then((promotions) => {
```

Comment:
If you could add the dispatch call for DATA_LOADING before this line that would be great!
```suggestion
      dispatch({ type: DispatchAction.DATA_LOADING });
      getPromotions([{ searchQuery: promotionListState.searchQuery }]).then((promotions) => {
```